crypto/x509/x509_vfy.c

```diff
@@ -471,16 +471,16 @@ static int check_chain(X509_STORE_CTX *ctx)
         int ret;
 
         x = sk_X509_value(ctx->chain, i);
-        CHECK_CB((ctx->param->flags & X509_V_FLAG_IGNORE_CRITICAL) == 0
-                 && (x->ex_flags & EXFLAG_CRITICAL) != 0,
-                 ctx, x, i, X509_V_ERR_UNHANDLED_CRITICAL_EXTENSION);
         CHECK_CB(!allow_proxy_certs && (x->ex_flags & EXFLAG_PROXY),
                  ctx, x, i, X509_V_ERR_PROXY_CERTIFICATES_NOT_ALLOWED);
+
         ret = X509_check_ca(x);
         switch (must_be_ca) {
         case -1:
             CHECK_CB((ctx->param->flags & X509_V_FLAG_X509_STRICT) != 0
-                         && ret != 1 && ret != 0,
+                     && ret != 1 && ret != 0 && ret != 4
```

Comment:
Why the addition of ret != 4 here? Shouldn't such certificate be actually rejected with strict check?

Reply:
> Only tangentially related, I wonder if we want to add named symbols for these various values instead of hardcoding magic numbers.

Definitely, from `crypto/x509/v3_purp.c`:
```c
/*-
 * CA checks common to all purposes
 * return codes:
 * 0 not a CA
 * 1 is a CA
 * 2 Only possible in older versions of OpenSSL when basicConstraints are absent
 *   new versions will not return this value. May be a CA
 * 3 basicConstraints absent but self-signed V1.
 * 4 basicConstraints absent but keyUsage present and keyCertSign asserted.
 * 5 Netscape specific CA Flags present
 */
```

Reply:
> Why the addition of ret != 4 here? Shouldn't such certificate be actually rejected with strict check?

I agree that this at least needs an explicit rationale.

Reply:
I had added this exception to make the following case succeed:
```
openssl verify -x509_strict -trusted self-signed_v3_EE_wrong_keyUsage.pem -partial_chain self-signed_v3_EE_wrong_keyUsage.pem
```
which is test case 31 - 
`strict verify allow self-signed_v3_EE_wrong_keyUsage.pem` in `25-test_req.t` 
but after your comment and thinking about it again I found that allowing certs of type
```
 * 4 basicConstraints absent but keyUsage present and keyCertSign asserted.
```
is not really adequate and is not the point of that use case.

The point rather is: if a cert is directly trusted and a self-issued EE cert, RFC 5280 does not apply, and neither should we check whether it is a CA cert or not or the like.
So I've move the exclusion of directly trusted certs before the "must_be_ca" checks.

Reply:
The last mentioned correction also made the other exception I had introduced:
```
&& ret != 3 /* accept self-signed v1 cert as CA */
```
obsolete, so I removed that as well.